flink-table/flink-table-planner/src/main/java/org/apache/flink/table/planner/plan/rules/logical/WrapJsonAggFunctionArgumentsRule.java

```diff
@@ -80,81 +81,90 @@ public WrapJsonAggFunctionArgumentsRule(Config config) {
+        final int newInputCount = inputCount + valueIndicesAfterProjection.size();
+        for (Integer jsonAggCallIndex : wrapIndicesMap.keySet()) {
+            final TargetMapping argsMapping =
+                    Mappings.create(MappingType.BIJECTION, newInputCount, newInputCount);
+            Integer valueIndex = wrapIndicesMap.get(jsonAggCallIndex);
+            argsMapping.set(valueIndex, valueIndicesAfterProjection.get(valueIndex));
+            final AggregateCall newAggregateCall =
+                    newWrappedArgCallList.get(jsonAggCallIndex).transform(argsMapping);
+            newWrappedArgCallList.set(jsonAggCallIndex, newAggregateCall);
         }
 
-        return aggCall.getArgList().stream().distinct().collect(Collectors.toList());
+        return aggregate.copy(
+                aggregate.getTraitSet(),
+                relBuilder.build(),
+                aggregate.getGroupSet(),
+                aggregate.getGroupSets(),
+                newWrappedArgCallList);
     }
 
     /**
-     * Adds (wrapped) projections for affected arguments of the aggregation.
+     * Adds (wrapped) projections for affected arguments of the aggregation. For duplicate
+     * projection fields, we only wrap them once and record the conversion relationship in the map
+     * valueIndicesAfterProjection.
      *
      * <p>Note that we cannot override any of the projections as a field may be used multiple times,
      * and in particular outside of the aggregation call. Therefore, we explicitly add the wrapped
      * projection as an additional one.
      */
     private void addProjections(
-            RelOptCluster cluster, RelBuilder relBuilder, List<Integer> affectedArgs) {
+            RelOptCluster cluster,
+            RelBuilder relBuilder,
+            List<Integer> affectedArgs,
+            int inputCount,
+            Map<Integer, Integer> valueIndicesAfterProjection) {
         final BridgingSqlFunction operandToStringOperator =
                 BridgingSqlFunction.of(cluster, JSON_STRING);
 
         final List<RexNode> projects = new ArrayList<>();
-        affectedArgs.stream()
-                .map(argIdx -> relBuilder.call(operandToStringOperator, relBuilder.field(argIdx)))
-                .forEach(projects::add);
-
-        relBuilder.projectPlus(projects);
-    }
-
-    /**
-     * Returns a {@link TargetMapping} that defines how the arguments of the aggregation must be
-     * mapped such that the wrapped arguments are used instead.
-     */
-    private TargetMapping getAggArgsMapping(int inputCount, List<Integer> affectedArgs) {
-        final int newCount = inputCount + affectedArgs.size();
-
-        final TargetMapping argsMapping =
-                Mappings.create(MappingType.BIJECTION, newCount, newCount);
-        for (int i = 0; i < affectedArgs.size(); i++) {
-            argsMapping.set(affectedArgs.get(i), inputCount + i);
+        int newProjectCount = 0;
+        for (Integer argIdx : affectedArgs) {
```

Comment:
we can change the param 'affectedArgs' to a `Set` and simplify the code further, e.g.,
```java
        for (Integer argIdx : affectedArgs) {
            valueIndicesAfterProjection.put(argIdx, inputCount + projects.size());
            projects.add(relBuilder.call(operandToStringOperator, relBuilder.field(argIdx)));
        }
```

Reply:
> we can change the param 'affectedArgs' to a `Set` and simplify the code further, e.g.,
> 
> ```java
>         for (Integer argIdx : affectedArgs) {
>             valueIndicesAfterProjection.put(argIdx, inputCount + projects.size());
>             projects.add(relBuilder.call(operandToStringOperator, relBuilder.field(argIdx)));
>         }
> ```

Done!